WordPress/gutenberg
# WP_Theme_JSON_Gutenberg: Add nested indexed array schema sanitization

## What?
- Add nested indexed array schema sanitization capabilities to the `WP_Theme_JSON_Gutenberg` class.
- Add settings.typography.fontFamilies validation in `WP_Theme_JSON_Gutenberg`
- Add unit tests.

This is an alternative approach to https://github.com/WordPress/gutenberg/pull/53273, and hopefully fixes: https://github.com/WordPress/gutenberg/issues/52798

Co-authrored by: @jffng 

## Why?
To sanitize data stored in indexed arrays, for example, font family definitions. Without this PR, that's not possible.

Currently, `WP_Theme_JSON` sanitization is not able to sanitize data contained on indexed arrays. So certain data from theme.json, for example, `settings.typography.fontFamilies` which is a JSON array, cannot be sanitized because when parsing the JSON data`WP_Theme_JSON` translates JSON arrays into PHP indexed arrays and the class is not capable of sanitizing that kind of data.





## How?
Treating the associative arrays and indexed arrays differently.

## Testing Instructions
1. Run the PHP unit tests:
`npm run test:unit:php`
2. Check that there are no failures.
----
1. Add the following snippet to a PHP file, for example, to your theme's `functions.php` file.
2. Check that the unwanted keys are removed from the output data.
```php
$theme_data = array(
  'version'  => '2',
  'badKey2' => 'I am Evil!!!!',
  'settings' => array(
    'badKey2' => 'I am Evil!!!!',
    'typography' => array(
      'badKey3' => 'I am Evil!!!!',
      'fontFamilies' => array(
        array (
            'badKey4' => 'I am Evil!!!!',
            'name'       => 'Piazzolla',
            'slug'       => 'piazzolla',
            'fontFamily' => 'Piazzolla',
            'fontFace'   => [
              array(
                'badKey5' => 'I am Evil!!!!',
                'fontFamily' => 'Piazzolla',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
              array(
                'badKey5' => 'I am Evil!!!!',
                'fontFamily' => 'Piazzolla',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
              array(
                'badKey5' => 'I am Evil!!!!',
                'fontFamily' => 'Piazzolla',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
            ],
        ),
        array (
            'badKey4' => 'I am Evil!!!!',
            'name'       => 'Inter',
            'slug'       => 'Inter',
            'fontFamily' => 'Inter',
            'fontFace'   => [
              array(
                'badKey5' => 'I am Evil!!!!',
                'fontFamily' => 'Inter',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
              array(
                'badKey5' => 'I am Evil!!!!',
                'fontFamily' => 'Inter',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
              array(
                'fontFamily' => 'Inter',
                'fontStyle'  => 'italic',
                'fontWeight' => '400',
                'src'        => 'https://example.com/font.ttf',
              ),
            ],
        )
      )
    ),
  ),
);
// Creates a new WP_Theme_JSON object with the new fonts to leverage sanitization and validation.
$theme_json = new WP_Theme_JSON_Gutenberg( $theme_data );
$data       = $theme_json->get_data();

echo('<pre>');
print_r($data);
echo('</pre>');
```
